File: YourBible/plugins/com-shortsands-utility/src/android/plugin/Utility.java
package plugin;

/**
 * Created by garygriswold on 1/9/18.
 */

import android.content.Context;
import android.content.res.Resources;
import android.provider.Settings;
import com.shortsands.utility.Sqlite3;
import java.util.ArrayList;
import java.util.Locale;
import org.apache.cordova.CordovaWebView;
import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CordovaInterface;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class Utility extends CordovaPlugin {

    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
	    if (action.equals("locale")) {
			Locale loc = Locale.getDefault();
			JSONArray message = new JSONArray();
			message.put(loc.toString());
			message.put(loc.getLanguage());
			message.put(loc.getScript());
			message.put(loc.getCountry());
			callbackContext.success(message);
	    } else if (action.equals("platform")) {
            callbackContext.success("Android");
        } else if (action.equals("modelType")) {
            callbackContext.success(android.os.Build.BRAND);
        } else if (action.equals("modelName")) {
			callbackContext.success(android.os.Build.MODEL);
        } else if (action.equals("open")) {
	        // what about threading each of these database calls
	        try {
		        Context context = this.cordova.getActivity();
	        	String dbname = args.getString(0);
				boolean copyIfAbsent = args.getBoolean(1);
				Sqlite3.openDB(context, dbname, copyIfAbsent);
				callbackContext.success();
			} catch(Exception error) {
				callbackContext.error(error.toString() + " on database open");
			}
	    } else if (action.equals("queryJS")) {
		    try {
		    	Sqlite3 database = Sqlite3.findDB(args.getString(0));
				String sql = args.getString(1);
				JSONArray values = args.getJSONArray(2);
				JSONArray resultSet = database.queryJS(sql, values);
				callbackContext.success(resultSet.toString());
		    } catch(Exception error) {
				callbackContext.error(error.toString() + " on database queryJS");
			}
		} else if (action.equals("executeJS")) {
			try {
				Sqlite3 database = Sqlite3.findDB(args.getString(0));
				String sql = args.getString(1);
				JSONArray values = args.getJSONArray(2);
				int rowCount = database.executeJS(sql, values);
				callbackContext.success(rowCount);
			} catch (Exception error) {
				callbackContext.error(error.toString() + " on database executeJS");
			}
		} else if (action.equals("bulkExecuteJS")) {
        	try {
				Sqlite3 database = Sqlite3.findDB(args.getString(0));
				String sql = args.getString(1);
				JSONArray values = args.getJSONArray(2);
				int rowCount = database.bulkExecuteJS(sql, values);
				callbackContext.success(rowCount);
			} catch (Exception error) {
				callbackContext.error(error.toString() + " on database bulkExecuteJS");
			}
		} else if (action.equals("close")) {
			try {
				Sqlite3 database = Sqlite3.findDB(args.getString(0));
				database.close();
				callbackContext.success();
			} catch (Exception error) {
				callbackContext.error(error.toString() + " on database close");
			}
		} else if (action.equals("listDB")) {
        	try {
        		ArrayList<String> files = Sqlite3.listDB(this.cordova.getActivity());
        		JSONArray json = new JSONArray(files);
        		callbackContext.success(json);
			} catch (Exception error) {
				callbackContext.error(error.toString() + " on database listDB");
			}
		} else if (action.equals("deleteDB")) {
			try {
				Sqlite3.deleteDB(this.cordova.getActivity(), args.getString(0));
				callbackContext.success();
			} catch (Exception error) {
				callbackContext.error(error.toString() + " on database listDB");
			}
        } else {
            return false;
        }
        return true;
    }
}

File: YourBible/plugins/com-shortsands-utility/src/ios/Utility.swift
//
//  Utility.swift
//  Utility
//
//  Created by Gary Griswold on 1/9/18.
//  Copyright © 2018 ShortSands. All rights reserved.
//

import Foundation
import UIKit

/*
 platform, modelType, modelName, deviceSize
 */

@objc(Utility) class Utility : CDVPlugin {
    
    @objc(platform:) func platform(command: CDVInvokedUrlCommand) {
        let message = "iOS"
        let result = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: message)
        self.commandDelegate!.send(result, callbackId: command.callbackId)
    }
    
    @objc(modelType:) func modelType(command: CDVInvokedUrlCommand) {
        let message = UIDevice.current.model
        let result = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: message)
        self.commandDelegate!.send(result, callbackId: command.callbackId)
    }
    
    @objc(modelName:) func modelName(command: CDVInvokedUrlCommand) {
        let message = DeviceSettings.modelName()
        let result = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: message)
        self.commandDelegate!.send(result, callbackId: command.callbackId)
    }
    
    @objc(deviceSize:) func deviceSize(command: CDVInvokedUrlCommand) {
        let message = DeviceSettings.deviceSize()
        let result = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: message)
        self.commandDelegate!.send(result, callbackId: command.callbackId)
    }
    
    @objc(open:) func open(command: CDVInvokedUrlCommand) {
	    DispatchQueue.global().sync {
		    do {
		    	let database = try Sqlite3.openDB(
		    				dbname: command.arguments[0] as? String ?? "",
		    				copyIfAbsent: command.arguments[1] as? Bool ?? false)
		    	let result = CDVPluginResult(status: CDVCommandStatus_OK)
                self.commandDelegate!.send(result, callbackId: command.callbackId)
		    } catch let err {
                self.returnError(error: err, command: command)
			}
		}
	}
	
    @objc(queryJS:) func queryJS(command: CDVInvokedUrlCommand) {
	    DispatchQueue.global().sync {
		    do {
			    let database = try Sqlite3.findDB(dbname: command.arguments[0] as? String ?? "")
			    let resultSet = try database.queryJS(
				    		sql: command.arguments[1] as? String ?? "",
				    		values: command.arguments[2] as? [Any?] ?? [])
                let json = String(data: resultSet, encoding: String.Encoding.utf8)
                let result = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: json)
                self.commandDelegate!.send(result, callbackId: command.callbackId)
			} catch let err {
                self.returnError(error: err, command: command)
			}
		}
	}
	
	@objc(executeJS:) func executeJS(command: CDVInvokedUrlCommand) {
		DispatchQueue.global().sync {
		    do {
			    let database = try Sqlite3.findDB(dbname: command.arguments[0] as? String ?? "")
			    let rowCount = try database.executeV1(
				    		sql: command.arguments[1] as? String ?? "",
				    		values: command.arguments[2] as? [Any?] ?? [])
                let result = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: rowCount)
                self.commandDelegate!.send(result, callbackId: command.callbackId)
			} catch let err {
				self.returnError(error: err, command: command)
			}
		}
	}
	
	@objc(close:) func close(command: CDVInvokedUrlCommand) {
		DispatchQueue.global().sync {
		    Sqlite3.closeDB(dbname: command.arguments[0] as? String ?? "")
			let result = CDVPluginResult(status: CDVCommandStatus_OK)
		    self.commandDelegate!.send(result, callbackId: command.callbackId)
	    }
	}
    
    @objc(listDB:) func listDB(command: CDVInvokedUrlCommand) {
        do {
            let files = try Sqlite3.listDB()
            let result = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: files)
            self.commandDelegate!.send(result, callbackId: command.callbackId)
        } catch let err {
            self.returnError(error: err, command: command)
        }
    }
    
    @objc(deleteDB:) func deleteDB(command: CDVInvokedUrlCommand) {
        do {
            try Sqlite3.deleteDB(dbname: command.arguments[0] as? String ?? "")
            let result = CDVPluginResult(status: CDVCommandStatus_OK)
            self.commandDelegate!.send(result, callbackId: command.callbackId)
        } catch let err {
            self.returnError(error: err, command: command)
        }
    }
    
    private func returnError(error: Error, command: CDVInvokedUrlCommand) {
        let message = Sqlite3.errorDescription(error: error)
        let result = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: message)
        self.commandDelegate!.send(result, callbackId: command.callbackId)
    }
}
